solr/core/src/test/org/apache/solr/core/TestConfigSetService.java

```diff
@@ -0,0 +1,126 @@
+  private static ZkTestServer zkServer;
+  private static SolrZkClient zkClient;
+
+  @BeforeClass
+  public static void startZkServer() throws Exception {
+    zkServer = new ZkTestServer(createTempDir("zkData"));
+    zkServer.run();
+    zkClient = new SolrZkClient(zkServer.getZkAddress("/solr"), 10000);
+  }
+
+  @AfterClass
+  public static void shutdownZkServer() throws IOException, InterruptedException {
+    zkClient.close();
+    if (null != zkServer) {
+      zkServer.shutdown();
+    }
+    zkServer = null;
+  }
+
+  public TestConfigSetService(Supplier<ConfigSetService> configSetService) {
+    this.configSetService = configSetService.get();
+  }
+
+  @ParametersFactory
+  @SuppressWarnings("rawtypes")
+  public static Iterable<Supplier[]> parameters() {
+    return Arrays.asList(
+        new Supplier[][] {
+          {() -> new ZkConfigSetService(zkClient)},
+          {() -> new FileSystemConfigSetService(createTempDir("configsets"))}
+        });
+  }
+
+  @Test
+  public void testConfigSetServiceOperations() throws IOException {
+    final String configName = "testconfig";
+    byte[] testdata = "test data".getBytes(StandardCharsets.UTF_8);
+
+    Path configDir = createTempDir("testconfig");
+    Files.write(configDir.resolve("file1"), testdata);
+    Files.createFile(configDir.resolve("file2"));
+    Files.createDirectory(configDir.resolve("subdir"));
+    Files.createFile(configDir.resolve("subdir").resolve("file3"));
+
+    configSetService.uploadConfig(configName, configDir);
+
+    assertTrue(configSetService.checkConfigExists(configName));
+    assertFalse(configSetService.checkConfigExists("dummyConfig"));
+
+    byte[] data = "file3 data".getBytes(StandardCharsets.UTF_8);
+    configSetService.uploadFileToConfig(configName, "subdir/file3", data, true);
+    assertArrayEquals(configSetService.downloadFileFromConfig(configName, "subdir/file3"), data);
+
+    data = "file4 data".getBytes(StandardCharsets.UTF_8);
+    configSetService.uploadFileToConfig(configName, "subdir/file4", data, true);
+    assertArrayEquals(configSetService.downloadFileFromConfig(configName, "subdir/file4"), data);
+
+    Map<String, Object> metadata = configSetService.getConfigMetadata(configName);
+    assertTrue(metadata.isEmpty());
+    metadata = configSetService.getConfigMetadata("dummyConfig");
+    assertTrue(metadata.isEmpty());
+
+    configSetService.setConfigMetadata(configName, Collections.singletonMap("trusted", true));
+    metadata = configSetService.getConfigMetadata(configName);
+    assertTrue(metadata.containsKey("trusted"));
+
+    configSetService.setConfigMetadata(configName, Collections.singletonMap("foo", true));
+    assertFalse(configSetService.getConfigMetadata(configName).containsKey("trusted"));
+    assertTrue(configSetService.getConfigMetadata(configName).containsKey("foo"));
+
+    List<String> configFiles = configSetService.getAllConfigFiles(configName);
+    assertTrue(configFiles.size() == 5);
```

Comment:
can you assert the list contents so we can see the expected strings?  I ask because this method is somewhat quirky / specific about the response, e.g. sorted & has intermediate dirs